FPS counter that does not need to stabilize after unpausing

danamlund opened the pull request:
This FPS counter only counts frames when the game is not paused.

This commit enables it in dwarfmonitor by default. But it is only shown if [FPS:YES] is set in init.txt.
Its positioned on top of the default FPS counter (when its pushed by the "*PAUSED*" text).

It does not support changing FPS:YES at runtime.

lethosor commented on plugins/lua/dwarfmonitor.lua:
```diff
@@ -146,6 +146,95 @@ function Widget_cursor:render_body(p)
+    if df.global.pause_state then
+        self.prev_paused = true
+    else
+        local clock = os.clock()
```
This is not portable - it returns the elapsed CPU time on Linux and macOS, not the absolute time, so I'm seeing stuff like "FPS: 342" and "FPS: 887" with a limit of 100.

lethosor commented on plugins/lua/dwarfmonitor.lua:
```diff
@@ -146,6 +146,95 @@ function Widget_cursor:render_body(p)
+    if df.global.pause_state then
+        self.prev_paused = true
+    else
+        local clock = os.clock()
```
This is not portable - it returns the elapsed CPU time on Linux and macOS, not the absolute time, so I'm seeing stuff like "FPS: 342" and "FPS: 887" with a limit of 100.

danamlund replied:
The counter is updated every second, and the value is the average of the latest five 1-second intervals of FPS. Without averaging, the FPS counter would jump unpredictable every time it was reset. I will fix it so the counter is always visible (but shows as "FPS:   -" the first second).

The counter seemed to work on my linux machine so I did not think to thoroughly understand os.clock(). 
To fix it I intend to add a variant of MiscUtils.cpp:getTimeMs64() to LuaApi.cpp:dfhack_module because lua does not provide millisecond precision timing by default. That way I can call dfhack.getTime() from lua that returns seconds since epoch with millisecond floating-point precision (similar to os.clock()).
Since lua stores numbers as 64 bit doubles, a 64 bit unsigned int will lose some precision in the year 200 million. I can live with this limitation.

And thank you very much for all your informative feedback as I get know the code base.

lethosor replied:
Oh, was it working on Linux? I was guessing you were testing on Windows because os.clock returns absolute time there, but I guess it would effectively return absolute time on Linux if you're hitting the FPS cap.

Your idea for wrapping getTimeMs64 is good. Lua 5.3 supports 64-bit integers, and we had to switch to that for 0.43.05, so it won't lose precision until it wraps around.

danamlund commented:
I have added the new method dfhack.getTimeMs(). And used that method instead of os.clock() to get accurate timing.
The counter is now algo accurate when the fps is limited by maxfps.
I also now show "FPS: ..." for the first second before enough information have been gathered to show the real FPS.

danamlund commented:
I just realized that this counter removes the graphical fps. 
And more general, I dont like that this counter replaces an existing df feature by default. 
I will rework it to show the new fps count number in addition to the old fps counters.

danamlund commented:
Turns out i did not need a new time method, dfhack.global.enabler.clock is the number of milliseconds df have been running.
This FPS counter now also show graphical fps, so its a complete replacement for the default fps counter.

I decided to show FPS as average over the last 3 seconds. Showing over 1 second made it jump between 40 and 70 on my medium fort. 3 seconds show a stable number that still change quickly.

The new FPS counter is only used in dwarf mode, in the main menu and adventurere the default fps counter is used.

I am thinking this FPS counter fits better as a tweak rather than a dwarfmonitor setting, but then you would not be able to change how many seconds you want average over (like we can currenty in dfhack-config/dwarfmonitor.json). What do you think?

When i play, FPS is an important resource, so I think a more responsive FPS counter is useful enough to enable by default (assuming [FPS:YES]). If this counter should not be default, I will undo the changes to dfhack-config/dwarfmonitor.json.

danamlund commented:
Rewrote new FPS counter as a tweak. Rebased commits to hide away the dwarfmonitor code.

This FPS counter look just like the default one except it does not increase when paused and it shows average over the last 3 seconds. 
The new FPS counter is only for fortress mode, and only when [FPS:YES] is set in init.txt.

The one bug I cannot figure out how to fix is that disabling fps counter through gui/settings-manager.lua will not work with this new FPS counter. You have to remove [FPS:YES] and restart dwarf fortress.

lethosor commented:
Looks good to me. This seems like more of a personal preference sort of tweak to me, while most of the tweaks in dfhack.init-example are bugfixes and small UI additions, so I don't think it should be in there for now. I'll make sure people know about it, though, especially pack maintainers, so we'll see what happens.

Anyway, thanks for your recent contributions! (and sorry for being slow to merge things sometimes.)

danamlund commented:
Sounds good
